src/layouts/base.astro

```diff
@@ -15,6 +16,7 @@ const { title } = Astro.props;
     <meta name="description" content="Astro description" />
     <meta name="viewport" content="width=device-width" />
     <link rel="icon" type="image/svg+xml" href="/favicon.svg" />
+    <link rel="canonical" href=`${Astro.site}${id}`/>
```

Comment:
I want the ULID versions of the pages to be the definitive source of truth. There will be potentially many vanity URLs to any one page, but there should only ever be a single canonical.